Mathlib/Condensed/Discrete/Colimit.lean

```diff
@@ -0,0 +1,373 @@
+/-
+Copyright (c) 2024 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.Condensed.Discrete.LocallyConstant
+import Mathlib.Condensed.Equivalence
+import Mathlib.Topology.Category.LightProfinite.Extend
+/-!
+
+# The condensed set given by left Kan extension from `FintypeCat` to `Profinite`.
+
+This file provides the necessary API to prove that a condensed set `X` is discrete if and only if
+for every profinite set `S = limᵢSᵢ`, `X(S) ≅ colimᵢX(Sᵢ)`, and the analogous result for light
+condensed sets.
+-/
+
+universe u
+
+noncomputable section
+
+open CategoryTheory Functor Limits FintypeCat CompHausLike.LocallyConstant
+
+attribute [local instance] ConcreteCategory.instFunLike
+
+namespace Condensed
+
+section LocallyConstantAsColimit
+
+variable {I : Type u} [Category.{u} I] [IsCofiltered I] {F : I ⥤ FintypeCat.{u}}
+  (c : Cone <| F ⋙ toProfinite) (X : Type (u+1))
+
+/-- The presheaf on `Profinite` of locally constant functions to `X`. -/
+abbrev locallyConstantPresheaf : Profinite.{u}ᵒᵖ ⥤ Type (u+1) :=
+  CompHausLike.LocallyConstant.functorToPresheaves.{u, u+1}.obj X
+
+namespace LocallyConstant
+
+/--
+The functor `locallyConstantPresheaf` takes cofiltered limits of finite sets with surjective
+projection maps to colimits.
+-/
+noncomputable def isColimitAux (hc : IsLimit c) [∀ i, Epi (c.π.app i)] :
+    IsColimit <| (locallyConstantPresheaf X).mapCocone c.op := by
+  refine Types.FilteredColimit.isColimitOf _ _ ?_ ?_
+  · intro (f : LocallyConstant c.pt X)
+    obtain ⟨j, h⟩ := Profinite.exists_locallyConstant.{_, u} c hc f
+    exact ⟨⟨j⟩, h⟩
+  · intro ⟨i⟩ ⟨j⟩ (fi : LocallyConstant _ _) (fj : LocallyConstant _ _)
+      (h : fi.comap (c.π.app i) = fj.comap (c.π.app j))
+    obtain ⟨k, ki, kj, _⟩ := IsCofilteredOrEmpty.cone_objs i j
+    refine ⟨⟨k⟩, ki.op, kj.op, ?_⟩
+    dsimp only [comp_obj, op_obj, functorToPresheaves_obj_obj, CompHausLike.coe_of,
+      Functor.comp_map, op_map, Quiver.Hom.unop_op, functorToPresheaves_obj_map]
+    -- Note: we might want to remove the `simps` attribute from `FintypeCat.toProfinite`; keeping
+    -- `toProfinite_obj` in the `dsimp` block above causes the following `ext` to fail.
+    ext x
+    obtain ⟨x, hx⟩ := ((Profinite.epi_iff_surjective (c.π.app k)).mp inferInstance) x
+    rw [← hx]
+    change fi ((c.π.app k ≫ (F ⋙ toProfinite).map _) x) =
+      fj ((c.π.app k ≫ (F ⋙ toProfinite).map _) x)
+    have h := LocallyConstant.congr_fun h x
+    rwa [c.w, c.w]
+
+/--
+The functor `locallyConstantPresheaf` takes a profinite set written as the cofiltered limit of its
+discrete quotients (`S.asLimit`) to the corresponding colimit.
+-/
+noncomputable def isColimit (S : Profinite) :
```

Comment:
Most of the `def`s in this PR do not come with `@[simps]` or other forms of equation lemmas/API.

Do you think `@[simps]` just doesn't work well enough? I'm a bit worried that these defn's will be be hard to work with if there are no supporting lemmas at all.

Reply:
My idea was that one shouldn't interact that much with the `def`s in this file, but rather use the `Condensed.IsDiscrete` typeclass introduced in #14027 and access the equivalent characterizations via the `tfae` statements introduced in that same PR.

But I'll go through this and see if there are some useful simp lemmas to be extracted.